File: jsonantt/web/value-format.mjs
/** Display-only values. Existing multipliers are applied before this formatter. */
export const VALUE_SCALES = {units:[1,''], thousands:[1000,'K'], millions:[1e6,'M'], billions:[1e9,'B']};
export function valueFormatActive(style, field) {
  if (['task','id','name','description','start','end','date','effective_start','effective_end','milestone_date','duration','not_before','offset','marker_size'].includes(field)) return false;
  const fields = style.value_fields ?? ['cost'];
  return (!fields.length || fields.includes(field)) && (style.value_prefix != null || style.value_suffix != null
    || (style.value_scale ?? 'units') !== 'units' || style.value_decimals != null);
}
export function validateValueFormat(style) {
  if (!Object.hasOwn(VALUE_SCALES, style.value_scale ?? 'units')) throw new Error('Value scale must be units, thousands, millions, or billions.');
  if (style.value_decimals != null && (!Number.isInteger(style.value_decimals) || style.value_decimals < 0 || style.value_decimals > 8)) throw new Error('Value decimals must be an integer from 0 to 8.');
  if (style.value_fields !== undefined && (!Array.isArray(style.value_fields) || style.value_fields.some(field=>typeof field !== 'string' || !field.trim()))) throw new Error('Value fields must be a list of field names.');
  for (const key of ['value_prefix','value_suffix']) if (style[key] != null && typeof style[key] !== 'string') throw new Error(`${key} must be a string.`);
}
export function valueAffixes(value) {
  if (typeof value !== 'string') return {};
  const match = /^\s*([^\d+\-.]*)([+\-]?(?:\d[\d,]*)(?:\.\d+)?)\s*([^\d]*)\s*$/.exec(value);
  return match ? {prefix:match[1].trim(), suffix:match[3].trim()} : {};
}
export function chartValueAffixes(chart, field) {
  for (const task of chart.flat) {
    const spec = valueAffixes(task.raw[field]);
    if (spec.prefix || spec.suffix) return spec;
  }
  return {};
}
export function formatValue(amount, style, field, spec = {}) {
  if (!valueFormatActive(style, field)) return null;
  const [divisor, unit] = VALUE_SCALES[style.value_scale ?? 'units'];
  const places = style.value_decimals ?? 2;
  const value = amount / divisor;
  const number = Math.abs(value).toLocaleString('en-US', {minimumFractionDigits:style.value_decimals ?? 0, maximumFractionDigits:places, roundingMode:'halfEven'});
  const prefix = style.value_prefix ?? spec.prefix ?? '';
  const suffix = style.value_suffix ?? spec.suffix ?? '';
  return `${value < 0 && Number(number.replaceAll(',','')) !== 0 ? '-' : ''}${prefix}${number}${unit}${suffix ? ` ${suffix}` : ''}`;
}
export function valueUnitLabel(style, field, spec = {}) {
  if (!valueFormatActive(style, field)) return '';
  return [style.value_prefix ?? spec.prefix, (style.value_scale && style.value_scale !== 'units') ? style.value_scale : '', style.value_suffix ?? spec.suffix].filter(Boolean).join(' ');
}

File: jsonantt/value_format.py
"""Opt-in display formatting; never changes source amounts or allocations."""
from decimal import Decimal, ROUND_HALF_EVEN

SCALES = {'units': (1, ''), 'thousands': (1000, 'K'), 'millions': (1000000, 'M'), 'billions': (1000000000, 'B')}


def value_format_active(style, field):
    if field in {'task', 'id', 'name', 'description', 'start', 'end', 'date', 'effective_start', 'effective_end', 'milestone_date', 'duration', 'not_before', 'offset', 'marker_size'}:
        return False
    fields = style.value_fields
    return (not fields or field in fields) and (
        style.value_prefix is not None or style.value_suffix is not None
        or style.value_scale != 'units' or style.value_decimals is not None)


def validate_value_format(style):
    if style.value_scale not in SCALES:
        raise ValueError('value_scale must be units, thousands, millions, or billions')
    if style.value_decimals is not None and (type(style.value_decimals) is not int or not 0 <= style.value_decimals <= 8):
        raise ValueError('value_decimals must be an integer from 0 to 8')
    if not isinstance(style.value_fields, list) or any(not isinstance(field, str) or not field.strip() for field in style.value_fields):
        raise ValueError('value_fields must be a list of field names')
    for value in (style.value_prefix, style.value_suffix):
        if value is not None and not isinstance(value, str):
            raise ValueError('value_prefix and value_suffix must be strings')


def format_value(amount, style, field, spec=None):
    """Return formatted text, or None to retain the existing renderer format."""
    if not value_format_active(style, field):
        return None
    spec = spec or {}
    divisor, unit = SCALES[style.value_scale]
    places = 2 if style.value_decimals is None else style.value_decimals
    value = (Decimal(str(amount)) / Decimal(divisor)).quantize(Decimal(1).scaleb(-places), rounding=ROUND_HALF_EVEN)
    prefix = style.value_prefix if style.value_prefix is not None else spec.get('prefix', '')
    suffix = style.value_suffix if style.value_suffix is not None else spec.get('suffix', '')
    text = f'{abs(value):,.{places}f}'
    if style.value_decimals is None and '.' in text:
        text = text.rstrip('0').rstrip('.')
    return f'{"-" if value < 0 else ""}{prefix}{text}{unit}{(" " + suffix) if suffix else ""}'


def value_unit_label(style, field, spec=None):
    if not value_format_active(style, field):
        return ''
    spec = spec or {}
    prefix = style.value_prefix if style.value_prefix is not None else spec.get('prefix', '')
    suffix = style.value_suffix if style.value_suffix is not None else spec.get('suffix', '')
    scale = '' if style.value_scale == 'units' else style.value_scale
    return ' '.join(value for value in (prefix, scale, suffix) if value)
